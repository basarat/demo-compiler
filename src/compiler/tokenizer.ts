import { Token } from './types';

export function tokenizer(input: string): Token[] {
  let cursorPos = 0;

  const tokens: Token[] = [];

  while (cursorPos < input.length) {
    let char = input[cursorPos]!;

    if (/\s/.test(char)) {
      cursorPos++; continue;
    }

    if (char === '(') {
      tokens.push({ type: 'OpenParenToken' });
      cursorPos++; continue;
    }

    if (char === ')') {
      tokens.push({ type: 'CloseParenToken' });
      cursorPos++; continue;
    }

    if (char === '+') {
      tokens.push({ type: 'PlusToken' });
      cursorPos++; continue;
    }

    if (char === '-') {
      tokens.push({ type: 'MinusToken' });
      cursorPos++; continue;
    }

    const NUMBERS = /[0-9]/;
    if (NUMBERS.test(char)) {
      let value = '';
      while (NUMBERS.test(char)) {
        value += char;
        char = input[++cursorPos]!;
      }
      tokens.push({ type: 'NumericLiteral', value: +value });
      continue;
    }

    const LETTERS = /[a-z]/i;
    if (LETTERS.test(char)) {
      let value = '';
      while (LETTERS.test(char)) {
        value += char;
        char = input[++cursorPos]!;
      }
      tokens.push({ type: 'Identifier', value });
      continue;
    }

    throw new SyntaxError(`Kya kar rha hai tu??...Unexpected token: "${char}"`);
  }

  return tokens;
}