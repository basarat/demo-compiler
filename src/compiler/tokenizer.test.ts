import { tokenizer } from './tokenizer';
test('it should pass', () => {
  expect(tokenizer('log(1+3-4)')).toMatchSnapshot();
});
