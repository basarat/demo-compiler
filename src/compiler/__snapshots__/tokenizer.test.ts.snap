// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should pass 1`] = `
Array [
  Object {
    "type": "Identifier",
    "value": "log",
  },
  Object {
    "type": "OpenParenToken",
  },
  Object {
    "type": "NumericLiteral",
    "value": "1",
  },
  Object {
    "type": "PlusToken",
  },
  Object {
    "type": "NumericLiteral",
    "value": "3",
  },
  Object {
    "type": "MinusToken",
  },
  Object {
    "type": "NumericLiteral",
    "value": "4",
  },
  Object {
    "type": "CloseParenToken",
  },
]
`;
